### ISSUE TITLE
Add unit tests

### ISSUE THREAD

**Issue by hikari-no-yume** [hikari-no-yume/touchHLE#53]
Currently the only testing done is, well, manually running Super Monkey Ball and trying to play it. This is a very effective integration test, but it's not going to scale. It would be really good to have some automated testing to complement it. For the moment, I'm only interested in unit testing (fsvo unit…).

There's no particularly good reason that I didn't write unit tests so far, except that this is a hobby project. :)

The easiest targets are leaf functions with no dependencies on `Environment` or other heavy things (`Mem`, `ObjC`…). For example, the path resolving algorithm inside `fs.rs` and the ADPCM decoder in `ima4.rs` would be easy to add tests for.

With a bit more work, we can test things that do have a dependency on `Environment`, but don't need a binary to be loaded or a window to be set up. We can add a special constructor to `Environment` for testing that skips these things, and possibly allocates less memory than normal. Then we would be able to write simple tests of the C string functions, `NSString`, etc.

I think I'm okay with not trying to have unit tests for anything related to binaries, windowing, audio output, OpenGL, and actual file I/O.

**Comment by ciciplusplus** [comment]
One idea for integration testing: use official Apple toolkit of the time (Xcode 3.0 for MacOS X 10.5, i guess) for building a test app targeting iPhone OS 2, then use it as canary testing for touchHLE

**Comment by hikari-no-yume** [comment]
Yeah, that wouldn't be a bad idea. I could also see what old homebrew (jailbreak-only) apps there were, maybe some could be run in touchHLE. Those might have licenses that let us redistribute them.

I already have a plan for how to do automated testing with real apps though. I should do it at some point.

**Comment by hikari-no-yume** [comment]
There is now one actual unit test, and it's in path handling code: 09c749c5f514a1cc73bb2ca296e14a4653d7f5a5. :p

**Comment by ciciplusplus** [comment]
FYI examples of open source iOS apps of 2009 era, both are released under GPL license:
- https://github.com/id-Software/Wolf3D-iOS
- https://github.com/id-Software/DOOM-iOS

**Comment by hikari-no-yume** [comment]
Thanks to @ciciplusplus's efforts in https://github.com/hikari-no-yume/touchHLE/pull/97, we now also have a way to write integration tests in C or Objective-C that will run under emulation. ^^

**Comment by TylerJaacks** [comment]
I know this isn't probably the best place to discuss this but how exactly does #97 work? So we have an app built with the iPhone OS 2.0 SDK and we are running in touchHLE and validate the result from qsort? Is that correct?

**Comment by ciciplusplus** [comment]
> I know this isn't probably the best place to discuss this but how exactly does #97 work? So we have an app built with the iPhone OS 2.0 SDK and we are running in touchHLE and validate the result from qsort? Is that correct?

yes, this is correct
